### DISCUSSION TITLE
zod instanceof class in inputs causing error.

### DISCUSSION THREAD

**Discussion by mlnj** [trpc/trpc#4052] [Q&A] [answered]
I had a question on how to use zod validation for inputs that can be serialized to a TypeScript class.

Here the endpoint takes in an input obj which is an instance of the class ABC.

```
class ABC {
  name: string;
  age: number;
}

const appRouter = router({
  greeting: publicProcedure
    .input(
      z.object({
        name: z.string().nullish(),
        obj: z.instanceof(ABC),
      })
    )
    .query(({ input }) => {
      return {
        text: `${JSON.stringify(input.obj)}`,
      };
    }),
});
```

Running this gives this error: 

```
TRPCClientError: [
  {
    "code": "custom",
    "message": "Input not instance of ABC",
    "fatal": true,
    "path": [
      "obj"
    ]
  }
]
```

What am I doing wrong?

**Comment by kamilogorek** [comment, marked as answer]
JSON serialization doesn't preserve the information about the serialized classes, thus it's not possible to use `instanceof` validator.
